### PR TITLE
Feature/solrprocess

### PR DESCRIPTION
**Pull request by bahadasx**
This PR addresses #8.  The parsexml.py script will now send each `DATA RECORD` from a specifc json file to Solr after processing that record.  The script is also set up now so that it can either use the parsed documents and send each one to Solr as it is processing it (prior to creating the json doc) or can use an existing JSON doc if one has already been created.

### CODE REVIEW

File: parsexml.py [hunk 1]
```diff
@@ -9,39 +9,82 @@
...
+#get field(date) and return in a proper iso format
+def format_date(field,x):
+    date = dateutil.parser.parse(x.pop(field))
+    dateiso = date.isoformat()+'Z'
+    return dateiso
+
 #this function contains the code for parsing the xml file
 #and writing the results out to a json file
 def processFile(fname):
+    type = os.path.splitext(fname)[1]
```
Comment (arowla) [inline parsexml.py]: `type` is a reserved word in Python, so you should rename this to `filetype`.

File: parsexml.py [hunk 2]
```diff
@@ -9,39 +9,82 @@
...
-                else:
-                    logging.error("PDF text file does not exist. JSON file creation skipped.")
-                    return
+            records = doc['main']['DATA_RECORD']
+            for x in records:
+                docid = x.get('DOCUMENT_IMAGE_ID',x.get('DOCUMENT_NM'))
+                if type == ".xml":
+                    txtfn = os.path.join(os.path.dirname(fn),'PDF_image',docid+'.txt')
+                    if os.path.isfile(txtfn):
+                        x['LAST_MODIFIED_TS'] = format_date('LAST_MODIFIED_TS',x)
```
Comment (arowla) [inline parsexml.py]: Just a stylistic thing... Usually we pass in the object/record as the first argument, then the more specific info after that, so I would reverse the argument order here, e.g.:

`format_date(x, 'LAST_MODIFIED_TS')`

File: parsexml.py [hunk 3]
```diff
@@ -9,39 +9,82 @@
...
             #transform output to json and save to file with same name
-            with open(fn,'w') as outfile:
-                json.dump(doc,outfile)
-                logging.info("-- Processing of XML file and creation of json file complete")
+            if type == ".xml":
+                with open(fn,'w') as outfile:
+                    json.dump(doc,outfile)
+                    logging.info("-- Processing of XML file and creation of json file complete")
+                    
     except IOError as e:
```
Comment (arowla) [inline parsexml.py]: Can we move this `except` way up, to only wrap the code really relevant to what it's handling errors for?

File: parsexml.py [hunk 4]
```diff
@@ -9,39 +9,82 @@
...
+                    else:
+                        logging.info("TXT file: "+docid+".txt  does not exist. JSON file creation skipped.")
+                        return
+                if (args.solr):
+                    jsontext = json.dumps(x)
+                    #Send content to Solr to update index
+                    jsontext = '{"add":{ "doc":'+jsontext+',"boost":1.0,"overwrite":true, "commitWithin": 1000}}'
+                    with open(os.path.join(os.path.dirname(fname),'jsoncomplete.txt'),'a+') as logfile:
+                        logfile.seek(0)
+                        if docid+"\n" in logfile:
```
Comment (arowla) [inline parsexml.py]: I don't think this is going to work the way you want it to. `logfile` is just the file handle, and to my knowledge, you can't search it directly for content. Also, the `seek(0)` above is probably going to make your log append to the beginning of the file, rather than the end.

File: parsexml.py [hunk 5]
```diff
@@ -9,39 +9,82 @@
...
+                            logging.info("-- file: "+docid+"  already processed by Solr")
+                            continue
+                        else:
+                            logging.info("-- Sending file: "+docid+" to Solr")
+                            url = "http://54.208.116.77:8983/solr/ptab/update"
+                            headers = {"Content-type" : "application/json"}
+                            r = requests.post(url, data=jsontext, headers=headers)
+                            rdict = json.loads(r.text)
```
Comment (arowla) [inline parsexml.py]: You can just do `r.json()`

File: parsexml.py [hunk 6]
```diff
@@ -9,39 +9,82 @@
...
+                            if status == 0:
+                                logging.info("-- Solr update for file: "+docid+" complete")
+                            else:
+                                logging.info("-- Solr error for doc: "+docid+" error: "+', '.join("{!s}={!r}".format(k,v) for (k,v) in rdict.items()))
+                            
             #transform output to json and save to file with same name
-            with open(fn,'w') as outfile:
-                json.dump(doc,outfile)
-                logging.info("-- Processing of XML file and creation of json file complete")
+            if type == ".xml":
```
Comment (arowla) [inline parsexml.py]: It's a little hard to follow why it's writing a JSON file way at the end here.
1. These sections should be broken out into functions, so the overall steps are easier to read.
2. The basic conversion and save should be done before the load into Solr, so if anything goes wrong during the load step, then at least you already have a saved file to start back up from.

Comment (bahadasx) [inline parsexml.py]: Thanks.  I will work on modularizing the code and streamlining the process.

File: parse_xml.py [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+    return '.'.join(seq)
+
+#get field(date) and return in a proper iso format
+def formatDate(field,x):
+    date = dateutil.parser.parse(x.pop(field))
+    dateiso = date.isoformat()+'Z'
+    return dateiso
+
+#send document to Solr for indexing
+def sendToSolr(core, json):
```
Comment (arowla) [inline parse_xml.py]: Nothing appears to be calling this.

File: parse_xml.py [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+        return s
+    except ValueError:
+        msg = "Not a valid date: '{0}'.".format(s)
+        raise argparse.ArgumentTypeError(msg)
+
+def processFile(filename):
+    fn = changeExt(filename,'json')
+    if os.path.isfile(os.path.abspath(fn)):
+        if (args.solr):
+            logging.info("-- XML file: "+filename+" already processed.  Skipping Solr process.")
```
Comment (arowla) [inline parse_xml.py:120]: Logic here in the `if/else` looks backward.

File: parse_xml.py [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+        else:
+            logging.info("-- XML file: "+filename+" already processed.  Sending to Solr for indexing.")
+            readJSON(fn)
```
Comment (arowla) [inline parse_xml.py:123]: Not actually sending to Solr.

File: parse_xml.py (hunk 8 above)
Comment (arowla) [inline parse_xml.py:120]: Ahh, I see why it looks backward. Because the arg is named in what looks like a positive way when it's actually a negative. The argument should be named `--skip-solr`, otherwise it is confusing.

File: parsexml.py (hunk 4 above)
Comment (bahadasx) [inline parsexml.py]: The reason I needed to add the seek(0) is because append mode `a+` behaves differently on Linux and OS X.
http://bugs.python.org/issue22651

From Linux:

```
a+ - Open for reading and appending (writing at end of file). The file is created if it does not exist. The initial file position
for reading is at the beginning of the file, but output is always appended to the end of the file.
```

From OS X:

```
a+ - Open for reading and writing. The file is created if it does not exist. The stream is positioned at the end of the 
file. Subsequent writes to the file will always end up at the then current end of file, irrespective of any intervening 
fseek(3) or similar.
```

I tested this in the python console too.  Even though `seek(0)` goes to the beginning of the file for reading, `f.write()` still defaults to the last line of the file.

**Comment by bahadasx** [comment]
@arowla This PR is now ready to be reviewed.  Thanks!

File: parse_xml.py [hunk 10]
```diff
@@ -0,0 +1,177 @@
...
+                        help="Process file(s) for specific date(s) - format YYYYMMDD",
+                        nargs='*',
+                        type=validDate
+                       )
+    parser.add_argument(
+                        "-s",
+                        "--skipsolr",
+                        required=False,
+                        help="Pass this flag to skip Solr processing",
+                        action='store_false'
```
Comment (arowla) [inline parse_xml.py]: Shouldn't this be `store_true`? Otherwise it's a double negative.

**Comment by bahadasx** [comment]
@arowla I fixed the issue you pointed out.  Thanks.
